### ISSUE TITLE
UBID design

### ISSUE THREAD

**Issue by kflemin** [SEED-platform/seed#3957]
plan the UBID work for Q3

some notes:

1. add in the SQL commands for UBID functionality
2. refactor UBID and deprecate ULID fields
	1. there can now be multiple UBIDs on a property. one will be labeled "main ubid" or something like that
	2. can have UBIDs on parcels
	3. new "UBID" model in db
		1. need to be able to manage UBIDs on my property
		2. which one do I want to promote as the "main ubid"
		3. import a set of buildings that have ubids, I want to match on ubids, how do I do that? (not a string match, probabilistic since we have multiple ubids. need a 'threshold' for matching ubids)

**Comment by RDmitchell** [comment]
@perryr16 -- should any of this be tested in the UI?

**Comment by perryr16** [comment]
I thought the new UBID UI work had been tested, but I defer to you.

**Comment by RDmitchell** [comment]
I haven't tested it, and don't really know what the changes are. Is it documented anywhere (a spec or something) what changes were made? 

Have the changes been deployed to dev1 or other server for testing?

**Comment by RDmitchell** [comment]
@perryr16 -- is there a spreadsheet or CSV file that has multiple ubids defined for one property that I can use to test?

**Comment by perryr16** [comment]
@RDmitchell I used these test files to test the 'preferred' ubid logic. Each spreadsheet has the same 2 properties with their ubid slightly changed. 
[tf_4029_2_0.xlsx](https://github.com/SEED-platform/seed/files/14842988/tf_4029_2_0.xlsx)
[tf_4029_2_1.xlsx](https://github.com/SEED-platform/seed/files/14842989/tf_4029_2_1.xlsx)
[tf_4029_2_4.xlsx](https://github.com/SEED-platform/seed/files/14842990/tf_4029_2_4.xlsx)
[tf_4029_2_5.xlsx](https://github.com/SEED-platform/seed/files/14842991/tf_4029_2_5.xlsx)
Let me know if you're looking for something different

**Comment by RDmitchell** [comment]
@perryr16 -- thanks ... these should work !

**Comment by RDmitchell** [comment]
instance: dev1
SHA: 
Org: LBNL 245
For the case above, where there were 2 UBIDs separated by a semicolon, the program just imported all the data and is showing it as one UBID. It didn't generated an error
![image](https://github.com/SEED-platform/seed/assets/6314950/32e8016d-9f18-4509-a422-cda117597084)

I think this is a pretty obscure use case, and I don't think we need to support this functionality.

Not sure why I made this test file with the 2 ubids in the same field -- maybe at some point we did support this. 

I will go on to test the files that @perryr16 provided which are in a better format, ie, multiple lines for the same building, but with similar UBIDs.

**Comment by RDmitchell** [comment]
Actually, it looks like Sentry did throw an error, but SEED just imported the data as is. 
![image](https://github.com/SEED-platform/seed/assets/6314950/1690d2a5-7e03-473f-ba27-7f32638beda0)

**Comment by perryr16** [comment]
Thanks for the thorough testing!

**Comment by RDmitchell** [comment]
I imported the spreadsheet shown above, but with the double UBID in one field, and the program brought in the UBIDs, but they are not being displayed in the Map view. 
![image](https://github.com/SEED-platform/seed/assets/6314950/157bfa53-05bd-4c20-b37b-969492d14b4e)

Looking up the UBID using the NREL tool, it shows up as a valid UBID
![image](https://github.com/SEED-platform/seed/assets/6314950/80381c74-5004-44ea-808e-3da106c1216e)

**Comment by perryr16** [comment]
Can you confirm that they are showing up on the inventory detail map view?
![Screenshot 2024-04-02 at 3 32 48 PM](https://github.com/SEED-platform/seed/assets/58446472/3c91cd2f-5517-4ae3-ba89-5f6e0632296a)

**Comment by RDmitchell** [comment]
@perryr16 -- yes they are showing up in the Inventory Detail page -- I hadn't even seen that before !!

![image](https://github.com/SEED-platform/seed/assets/6314950/ac74fe48-7608-43f3-b698-e0483510c586)

**Comment by RDmitchell** [comment]
@perryr16 -- when you tested importing your files, what was the UBID checking value in Settings? 

I guess 1 means they have to be an exact match? 

So is 0.5 a good default? I have no idea what to set it to.

**Comment by perryr16** [comment]
Yeah, 1 is a perfect match, 0 is no match. If you want to get a better understanding of how ubids overlap, you can use the "compare ubids" action from the inventory list and make some changes to the ubids (I usually change some of the last four numbers separated by "-")

I'm not sure what users would consider a 'match', but I think 0.5 would be an ok default. 
![Screenshot 2024-04-02 at 3 49 46 PM](https://github.com/SEED-platform/seed/assets/58446472/c574828e-9efb-4ea6-ac3d-b1c1b883e90b)

**Comment by RDmitchell** [comment]
@perryr16 -- where do I get to the Compare UBIDs screen?

**Comment by RDmitchell** [comment]
@perryr16 -- should I make a new issue for the fact that the UBIDs are showing up in the detail but not the list view of the Inventory?

**Comment by RDmitchell** [comment]
@perryr16 -- it is still a bit unclear to me what the outcome should be for each of your test cases. Maybe we can talk about it tomorrow on the iteration call.

**Comment by perryr16** [comment]
- The compare ubid screen is from Inventory List Actions dropdown.
- Are you referring to the test procedure laid out in this PR? https://github.com/SEED-platform/seed/pull/4074.

**Comment by perryr16** [comment]
@RDmitchell here is a more detailed testing procedure

- upload a few properties with ubids - [tf_4029_2_0.xlsx](https://github.com/SEED-platform/seed/files/14842988/tf_4029_2_0.xlsx)
- set org setting ubid threshold to something small like 0.1 to allow matching
- upload a same properties with slightly different UBIDs (ex: 85FPPRR9+3C-0-0-0-0 ---> 85FPPRR9+3C-0-0-0-1) - [tf_4029_2_1.xlsx](https://github.com/SEED-platform/seed/files/14842989/tf_4029_2_1.xlsx)
- properties and ubids should have been merged with the preferred set to the incoming
- view Inventory Detail UBIDs page, verify map
- change the preferred UBID
- map should update
- check ubid CRUD - from the inventory detail UBIDs page create a UBID, delete a UBID
- set org setting ubid threshold to something high like 0.9 to prevent matching
- upload the same properties with different UBIDS (ex: 85FPPRR9+3C-0-0-0-1 ---> 85FPPRR9+3C-0-0-0-5) - [tf_4029_2_5.xlsx](https://github.com/SEED-platform/seed/files/14842991/tf_4029_2_5.xlsx)
- property list should have created new properties as the ubid threshold was not met
- set org setting ubid threshold to something small like 0.1
- upload same properties with different UBIDs - [tf_4029_2_4.xlsx](https://github.com/SEED-platform/seed/files/14842990/tf_4029_2_4.xlsx)
- existing properties should have been merged into the incoming properties
- ensure ubid history is preserved on Inventory Detail UBIDs
- merge 2 properties together, ensure ubid history
- unmerge property, ubid history should revert to pre-merge history

**Comment by RDmitchell** [comment]
instance: dev1
SHA: 
Org: LBNL 245

Here are my results

1. uploaded tf_4029_2_0, resulted in 3 properties
2. Set criteria to 0.1
![image](https://github.com/SEED-platform/seed/assets/6314950/e2f1295c-a20e-4304-9184-ddc111ee435f)
3. uploaded tf_4029_2_1
4. Properties merged and preferred set to incoming UBID
 ![image](https://github.com/SEED-platform/seed/assets/6314950/caf361de-7066-47c4-a04a-b02b80b50205)
**Results for Property 1001, NREL CAFE**
![image](https://github.com/SEED-platform/seed/assets/6314950/0b5a4cad-12ce-459e-99f3-521a91dd8513)
Detail View
![image](https://github.com/SEED-platform/seed/assets/6314950/e0bf52cb-bf39-4782-bc17-bf0cbc5b01c0)
5. UBID Detail View - preferred set to incoming record UBID
![image](https://github.com/SEED-platform/seed/assets/6314950/283207b7-f39f-4dc8-a95f-02bd9e9c0690)
6. Changed the preferred UBID to the first one
7. Map updated
![image](https://github.com/SEED-platform/seed/assets/6314950/6b445980-290c-4f73-8aed-67d2e39379e2)
* **Results for Property 1002, NREL FTLB**
![image](https://github.com/SEED-platform/seed/assets/6314950/d6a18fa2-67a1-4055-80ba-0d946df04075)
Default Preferred (most recent addition)
![image](https://github.com/SEED-platform/seed/assets/6314950/437dca43-9f22-43e3-b119-f567b1f9eeff)
Changed to the one with just one box
![image](https://github.com/SEED-platform/seed/assets/6314950/666d6b5d-a47c-473f-b0fd-fa18a12498dd)

**Comment by perryr16** [comment]
@RDmitchell, your process is correct. I'm not sure what's going on with the network activity, I'll look into it. 

The intent is that SEED will detect the 2 properties that are a match to the incoming property and merge all 3 together

**Comment by RDmitchell** [comment]
@perryr16 -- I cleaned out the inventory for my org (LBNL 245) and the reimported file #1. I then imported file #2, per the steps above, keeping the threshold at 0.1, and the program has been stuck in the state below for more than 30 minutes. 

No errors, just stuck on step 4/6 -- do you think this might be a caching issue? 
![image](https://github.com/SEED-platform/seed/assets/6314950/96196e42-3a78-454d-a623-80e47179dbec)

I will start over and keep the developer tools open, which has "Disable cache" checked.

**Comment by RDmitchell** [comment]
@perryr16 -- having developer tools open doesn't seem to have helped. Still on 1% of Matching Data (4/6): Merging State Pairs after 4 minutes. 

Do  you want me to add you to this org and see if you can replicate this? Maybe it's something on my end. Seems like my network has been very slow these past few days.

**Comment by perryr16** [comment]
Yeah, please add me. It's strange that you're getting different behavior the second time around.

**Comment by RDmitchell** [comment]
@perryr16 -- I just sent you an invite to the org. 

Yes, definitely weird. I bet it's something on my end!

**Comment by perryr16** [comment]
Same result, stuck on 4/6. I'll try locally and on dev2 and get back to you

**Comment by RDmitchell** [comment]
I'm glad you can reproduce the problem !! That's always the first step ! Thanks !

**Comment by perryr16** [comment]
I successfully tested on dev2 with a cycle that has 16000, although it did hang on 1% for a second, then finished. I did notice an issue preserving access levels at step 14

https://github.com/SEED-platform/seed/assets/58446472/ac7e2ded-1426-41be-9106-18f07a1947c6

dev1 is only 1 commit behind dev2 and that commit wouldnt be causing this issue. I'm not sure whats causing the hold up

**Comment by RDmitchell** [comment]
@perryr16 -- should I test on dev2 for this and then close the issue?

**Comment by RDmitchell** [comment]
close it if everything works as intended, that is (!)

**Comment by perryr16** [comment]
Yeah, that works for me!

**Comment by RDmitchell** [comment]
@perryr16 -- am trying the same thing on dev2 and I got further this time, but it is now hanging, in the same way, ie, Matching Data (4/6) for the last file to be imported, ie, tf_4029_2_4.xlsx.

I left the duplicate properties from the previous import -- I wonder if that is causing a problem.

Should I try deleting the duplicate, non-merged files from the 2_5 import?

**Comment by RDmitchell** [comment]
@perryr16 -- well maybe I'm just not patient enough. I waited 15 minutes then dismissed the merge.

But when I look at the inventory now, it looks like all the records were merged into one ??

![image](https://github.com/SEED-platform/seed/assets/6314950/e485bb58-4bfc-442b-8e0d-0e8599d3eca4)

![image](https://github.com/SEED-platform/seed/assets/6314950/bb3d75e4-4eda-467c-aeac-8858fa802b98)

**Comment by perryr16** [comment]
curious! thanks I'll try it out

**Comment by perryr16** [comment]
I also got hung on 4/6 on dev2 in org LBNL 245. I think you can put this issue on hold until I figure out what's causing the hangup.

**Comment by RDmitchell** [comment]
OK, I will add a "testing question" label. thx

**Comment by perryr16** [comment]
@RDmitchell, I believe I've gotten to the bottom of this.

So there were a few things on my end that led to the issue you saw. First, the ubids for the Cafe and FTLB in files _4, and _5 were located too close and had overlap with all other ubids. I didn't notice it beucase I never tested with ubid as the sole matching criteria (I had PM property ID as a second matching field which covered up the error scenario). Long story short, if ubid is the sole matching criteria and all properties have some overlap the app merges the first property with all existing properties, then the second property is stranded and the app hangs.

I imagine this fix will make it onto dev 2 in the next 2 weeks. Here are some updated test files with more well-defined ubids.
[2props_ubids_0.xlsx](https://github.com/SEED-platform/seed/files/14890882/2props_ubids_0.xlsx)
[2props_ubids_1.xlsx](https://github.com/SEED-platform/seed/files/14890883/2props_ubids_1.xlsx)
[2props_ubids_4.xlsx](https://github.com/SEED-platform/seed/files/14890884/2props_ubids_4.xlsx)
[2props_ubids_5.xlsx](https://github.com/SEED-platform/seed/files/14890885/2props_ubids_5.xlsx)
